### DISCUSSION TITLE
how can I do "extract method" on the block of code in visual mode?

### DISCUSSION THREAD

**Discussion by MorKaragh** [mfussenegger/nvim-jdtls#694] [Q&A]
Hi!

It looks like "extract method" refactoring has quite a limited capabilities, but I am not sure if I have everything set up correctly.

Imagine for example that we are going to extract the whole method body into other method (a useless move in reality, but good as an example here). 

Is jdtls capable of doing such an extraction? Or is it suitable only for simple extractions like one single statement (which is working fine) ?

**Comment by mfussenegger** [comment]
You can call `code_action` or the `extract_method` function in visual mode to specify the range you want to extract.

E.g with keymaps like these.:

```lua
vim.keymap.set('v', 'crm', [[<ESC><CMD>lua require('jdtls').extract_method(true)<CR>]], { silent = true })
vim.keymap.set({"n", "v"}, "<a-CR>", vim.lsp.buf.code_action, {}))
```
